Add to controller framework; use in scheduler

lavalamp opened the pull request:
Fixes #6524

lavalamp committed:
add informational output to test

lavalamp committed:
Add easy setup for simple controller

Also add tests; coverage up to 86.7%

lavalamp committed:
Make scheduler not miss deletion events even in the case of a resync.

thockin commented on pkg/controller/framework/controller_test.go:
```diff
@@ -113,3 +113,60 @@ func Example() {
+
+	// Run the controller and run it until we close stop.
+	stop := make(chan struct{})
+	controller.Run(stop)
```
Shouldn't this be wrapped in a goroutine?

lavalamp replied:
Run itself starts two goroutines.

thockin replied:
I figured.  My point was that, as I understand, the general Go pattern is
for the caller to create the goroutine, rather than a library function.
It's all perspective, i guess, but as a consumer of this library I would
expect to say `go controller.Run(stop)`  Not a huge deal.

On Tue, Apr 7, 2015 at 5:16 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/controller/framework/controller_test.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6546#discussion_r27933403>
> :
>
> > +				key, err := framework.DeletionHandlingMetaNamespaceKeyFunc(obj)
> > +				if err != nil {
> > +					key = "oops something went wrong with the key"
> > +				}
> > +
> > +				// Record some output when items are deleted.
> > +				outputSetLock.Lock()
> > +				defer outputSetLock.Unlock()
> > +				outputSet.Insert(key)
> > +			},
> > +		},
> > +	)
> > +
> > +	// Run the controller and run it until we close stop.
> > +	stop := make(chan struct{})
> > +	controller.Run(stop)
>
> Run itself starts two goroutines.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6546/files#r27933403>
> .
>

lavalamp replied:
Yeah, I don't have a strong opinion, but I wasn't aware of that pattern and I think we're (I am) breaking it all over the place... hm ok I can change this.

thockin commented:
Looks nice.

lavalamp commented:
shippable is a flake, I filed #6548

lavalamp committed:
switch to require 'go controller.Run()'

lavalamp commented:
Hm, I want to extend the testing a bit before this merges. Will update this tomorrow.

lavalamp commented:
OK, this is ready to go. @davidopp

lavalamp commented:
Shippable is an unrelated flake.

lavalamp commented:
Rebased

derekwaynecarr commented:
Subscribe so I can track this merge to know when to update namespace controller so deletion is watch based and not just periodic synch

davidopp commented on pkg/client/cache/store.go:
```diff
@@ -74,6 +77,10 @@ func MetaNamespaceKeyFunc(obj interface{}) (string, error) {
 
+// ExplicitKey can be passed to MetaNamespaceKeyFunc if you have the key for
+// the object but not the object itself.
+type ExplicitKey string
```
Move this above definition of MetaNamespaceKeyFunc() to improve readability?

davidopp commented on pkg/controller/framework/controller.go:
```diff
@@ -102,3 +103,117 @@ func (c *Controller) processLoop() {
+
+// ResourceEventHandler can handle notifications for events that happen to a resource.
+// The events are informational only, so you can't return an error.
+// OnDelete will get the final state of the item if it is known, otherwise it will get
```
If you're going to describe the semantics of OnDelete (which I think is good), then I would suggest also doing the same for OnAdd() and OnUpdate(). BTW, is there some way OnUpdate() could get a copy of both the old and new value, instead of just the new value?

lavalamp replied:
adding description

OnUpdate could be extended to do that, but we can't guarantee that every single change will cause a distinct OnUpdate call-- sometimes changes will be aggregated (because we could have to re-list).

davidopp commented on pkg/controller/framework/controller.go:
```diff
@@ -102,3 +103,117 @@ func (c *Controller) processLoop() {
+// ResourceEventHandlerFuncs is an adaptor to let you easily specify as many
+// notification functions as you want while still implementing
+// ResourceEventHandler.
+type ResourceEventHandlerFuncs struct {
```
I didn't understand the part about "as many notification functions as you want." It looks to me like you specify at most one Add, one Update, and one Delete function?

davidopp commented on pkg/controller/framework/controller.go:
```diff
@@ -102,3 +103,117 @@ func (c *Controller) processLoop() {
+	return cache.MetaNamespaceKeyFunc(obj)
+}
+
+// NewInformer returns a cache.Store and a controller for populating the store
```
This is a great explanation of the return values but can you also explain the arguments?

davidopp commented on pkg/controller/framework/controller.go:
```diff
@@ -102,3 +103,117 @@ func (c *Controller) processLoop() {
+	resyncPeriod time.Duration,
+	h ResourceEventHandler,
+) (cache.Store, *Controller) {
+	// This will hold the downstream state, as we know it.
```
Maybe "client" instead of "downstream"?

davidopp commented on pkg/controller/framework/controller_test.go:
```diff
@@ -113,3 +116,157 @@ func Example() {
+func TestHammerController(t *testing.T) {
+	// This test executes a bunch of requests through the fake source and
+	// controller framework to make sure there's no locking/threading
+	// errors.
```
What will the test do if there are locking/threading errors?

lavalamp replied:
Hang forever. Or trigger the race detector.

davidopp replied:
Please add that to the comment.

davidopp replied:
Ping.

davidopp commented on pkg/controller/framework/controller_test.go:
```diff
@@ -113,3 +116,157 @@ func Example() {
+		outputSet[key] = append(outputSet[key], eventType)
+	}
+
+	// Make a controller that immediately deletes anything added to it, and
```
Is this comment correct? I agree the controller deletes anything added to it, but it also seems to do the same for things updated and deleted.

lavalamp replied:
Sorry copy pasta error

davidopp commented on pkg/controller/framework/controller_test.go:
```diff
@@ -113,3 +116,157 @@ func Example() {
+				pod.ObjectMeta.Namespace = "default"
+				// Add, update, or delete randomly.
+				// Note that these pods are not valid-- the fake source doesn't
+				// call validation or anything.
```
either you meant "for anything" or the phrase is not adding any information.

davidopp commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -59,13 +64,40 @@ func NewConfigFactory(client *client.Client) *ConfigFactory {
+	c.ScheduledPodLister.Store, c.scheduledPodPopulator = framework.NewInformer(
+		c.createAssignedPodLW(),
+		&api.Pod{},
+		time.Second*30,
```
You have some hard-coded constants in this file. Can you define them as const variables instead? This will make it easier to understand their purpose (if you choose good variable names) and to change them.

lavalamp replied:
sorry-- that was a rebase error, remnants of my etcd patch from last week.

lavalamp commented:
Comments addressed, rebased, some commits squashed-- PTAL

davidopp commented on pkg/controller/framework/controller.go:
```diff
@@ -102,3 +103,130 @@ func (c *Controller) processLoop() {
+//  * lw is list and watch functions for the source of the resource you want to
+//    be informed of.
+//  * objType is an object of the type that you expect to receieve.
+//  * resyncPeriod: if non-zero, will re-list this often (you will get OnUpdate
```
I don't understand "you will get OnUpdate calls, even if nothing changed" Or more to the point, I guess I understand it but it seems sub-optimal, especially since you don't pass both the old and new object to OnUpdate(), so the function receiving the callback may not know whether it changed (or do you assume the function receiving the callback always has a copy of the old version?).

lavalamp replied:
I will change OnUpdate to pass the old object as we knew it.

davidopp replied:
Great, thanks!

davidopp commented:
Just a couple of remaining nits.

BTW, after you have submitted everything for controller framework, can you write up a doc explaining how to use it? The comments are good but I think having a HOWTO kind of thing would help people who want to write their own controllers.

lavalamp commented:
Comments addressed-- I changed the update function, since it's not hard and seems useful. Also added another test. PTAL-- this is blocking a number of others, so I'd like to fix further comments in a followup PR if you don't want anything major changed.

lavalamp commented:
And yes, I can write up a doc. BTW, the Example() functions will show up in godoc.

davidopp commented on pkg/controller/framework/controller.go:
```diff
@@ -102,3 +103,126 @@ func (c *Controller) processLoop() {
+//      change. OnUpdate is also called when a re-list happens, and it will
+//      get called even if nothing changed. This is useful for periodically
+//      evaluating or syncing something.
+//  * OnDelete will get the final state of the item if it is known, otherwise
```
In a followup PR, it might be good to add some detail to this comment explaining under what circumstances the final state would be known and when it wouldn't. That will held a client know when it can depend on getting this.

davidopp commented on pkg/controller/framework/controller_test.go:
```diff
@@ -113,3 +116,272 @@ func Example() {
+	outputSetLock := sync.Mutex{}
+	outputSet := util.StringSet{}
+
+	// Make a controller that immediately deletes anything added to it, and
```
I take it that the call to source.Delete() in AddFunc() will trigger DeleteFunc() ?

lavalamp replied:
Non-synchronously, yes.

davidopp commented on pkg/controller/framework/controller_test.go:
```diff
@@ -113,3 +116,272 @@ func Example() {
+	for _, key := range outputSet.List() {
+		fmt.Println(key)
+	}
+	// Output:
```
Can you make it test the result of outputSet.List() for what you expect, rather than just writing it in a comment?

lavalamp replied:
Please read the go doc about how example functions work: http://golang.org/pkg/testing/ (search for "Examples")

davidopp commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -144,39 +161,25 @@ func (f *ConfigFactory) CreateFromKeys(predicateKeys, priorityKeys util.StringSe
 		// Disable this code until minions support watches. Note when this code is enabled,
 		// we need to make sure minion ListWatcher has proper FieldSelector.
-		cache.NewReflector(f.createMinionLW(), &api.Node{}, f.NodeLister.Store, 0).Run()
+		cache.NewReflector(f.createMinionLW(), &api.Node{}, f.NodeLister.Store, 10*time.Second).RunUntil(f.StopEverything)
```
In a followup CL can you make the 10*time.Second a flag?

davidopp commented on pkg/controller/framework/controller_test.go:
```diff
@@ -113,3 +116,272 @@ func Example() {
+
+	var testDoneWG sync.WaitGroup
+
+	// Make a controller that immediately deletes anything added to it, and
```
This comment is not correct.

davidopp commented on pkg/controller/framework/controller_test.go:
```diff
@@ -113,3 +116,272 @@ func Example() {
+			defer wg.Done()
+			testDoneWG.Add(1)
+			name = "c-" + name
+			source.AddDropWatch(pod(name, FROM))
```
What is the purpose of AddDropWatch? IIUC you don't have a ResourceEventHandlerFunc handler for Add anyway.

lavalamp replied:
That doesn't matter. The Add populates the cache which the thing that calls update compares against. Dropping the watch means that the subsequent Update will be perceived by the client as an add.

davidopp commented on pkg/controller/framework/controller_test.go:
```diff
@@ -113,3 +116,272 @@ func Example() {
+			testDoneWG.Add(1)
+			name = "b-" + name
+			source.Add(pod(name, FROM))
+			source.ModifyDropWatch(pod(name, TO))
```
Can you verify my understanding of the intent of the *DropWatch() functions? The corresponding ResourceEventHandlerFuncs function doesn't get called (for example, ModifyDropWatch() won't cause UpdateFunc to be invoked), but because the set is still mutated, the next time a Modify() is invoked, a change made by a previous *DropWatch() function can show up as the source for a transition?

lavalamp replied:
No, there's a layer in the client that compares stuff received over the watch with a local cache and generates a call to add/update/delete. It is this layer that we are withholding the watch notification from. Basically we're purposely causing the client's internal state to become decoupled from the server's state, to demonstrate that the client does expected things if that happens.

davidopp commented:
I made some comments/questions/suggestions, none of them are blockers for merging but it seems that Travis is failing on v1beta3 with a race condition related to this PR, so you should probably look into that before we merge.

lavalamp commented:
Actually this race is possible without this PR. I have a fix, but it relies on a PR that hasn't merged yet.

davidopp commented:
If you think this PR doesn't make things worse, feel free to merge.

lavalamp committed:
fix race

lavalamp commented:
OK, I've fixed the race.

lavalamp commented:
Hm, still another test problem, it's too slow... looking at it.

lavalamp commented:
OK, let's see if travis still chokes on this.

lavalamp commented:
OK, I will add the comments you requested in a follow-up.

smarterclayton commented on pkg/conversion/deep_copy.go:
```diff
@@ -17,22 +17,27 @@ limitations under the License.
+	v := reflect.New(reflect.TypeOf(source))
+
+	buff := &bytes.Buffer{}
+	enc := gob.NewEncoder(buff)
```
Creating an encoder on the fly makes this much more expensive than it has to be, because all the reflections have to be recopied.  The encoder should be static (assuming we have a finite number of types) and that will make this much less expensive.
